module/zfs/mmp.c

```diff
@@ -0,0 +1,316 @@
+ * there is a problem, and a new write will also block).
+ *
+ * We try 10 times to pick a random leaf without an outstanding write.
+ * If 90% of the leaves have pending writes, this gives us a >65%
+ * chance of finding one we can write to.  There will be at least
+ * (zfs_mmp_fail_intervals) tries before the inability to write an MMP
+ * block causes serious problems.
+ */
+
+vdev_t *
+vdev_random_leaf(spa_t *spa)
+{
+	vdev_t *vd, *child;
+	int pending_writes = 10;
+
+	ASSERT(spa);
+	ASSERT(spa_config_held(spa, SCL_STATE, RW_READER) == SCL_STATE);
+
+	/*
+	 * Since we hold SCL_STATE, neither pool nor vdev state can
+	 * change.  Therefore, if the root is not dead, there is a
+	 * child that is not dead, and so on down to a leaf.
+	 */
+
+	if (vdev_is_dead(spa->spa_root_vdev))
+		return (NULL);
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+
+		if (vdev_is_dead(child))
+			continue;
+
+		if (child->vdev_ops->vdev_op_leaf &&
+		    child->vdev_mmp_pending) {
+			if (pending_writes-- > 0)
+				continue;
+			else
+				return (NULL);
+		}
+
+		vd = child;
+	}
+	return (vd);
+}
+
+static void
+mmp_write_done(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	mmp_thread_state_t *mts = zio->io_private;
+
+	mutex_enter(&mts->mmp_io_lock);
+
+	vd->vdev_mmp_pending = 0;
+	if (zio->io_error == 0)
+	{
+		mts->mmp_delay = gethrtime() - mts->mmp_last_write;
+		mts->mmp_last_write = gethrtime();
+	}
+
+	mutex_exit(&mts->mmp_io_lock);
+}
+
+/*
+ * Choose a random vdev, label, and MMP block, and write over it
+ * with a copy of the last-synced uberblock, whose timestamp
+ * has been updated to reflect that the pool is in use.  Use provided
+ * char *buf and abd_t *abd to avoid repeatedly allocating them.
+ */
+static void
+mmp_write_uberblock(dsl_pool_t *dp, abd_t *abd, char *buf, int bufsize)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	vdev_t *vd;
+	zio_t *zio;
+	int l, n;
+	int flags = ZIO_FLAG_CONFIG_WRITER | ZIO_FLAG_CANFAIL;
+
+	/* Copy the latest uberblock and update the time */
+	(void) memcpy(buf, &spa->spa_ubsync, sizeof (uberblock_t));
+	((uberblock_t *)buf)->ub_timestamp = gethrestime_sec();
+	abd_copy_from_buf(abd, buf, bufsize);
+
+	spa_config_enter(spa, SCL_STATE, FTAG, RW_READER);
+	vd = vdev_random_leaf(spa);
+	if (vd) {
+		n = spa_get_random(MMP_BLOCKS_PER_LABEL);
+		l = spa_get_random(VDEV_LABELS);
+		zio = zio_root(spa, NULL, NULL, flags);
+		vd->vdev_mmp_pending = gethrtime();
+		vdev_mmp_write(zio, vd, l, n, abd, mmp_write_done,
+		    mmp, flags | ZIO_FLAG_DONT_PROPAGATE);
+		zio_nowait(zio);
+	}
+	spa_config_exit(spa, SCL_STATE, FTAG);
+}
+
+static void
+mmp_thread(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	abd_t *abd;
+	char *buf;
+	int bufsize = (1ULL << MAX_UBERBLOCK_SHIFT);
+	uint last_zfs_mmp_interval = zfs_mmp_interval;
```

Comment:
You can declare this as `uint64_t` for the local and use `ulong` for `zfs_mmp_interval` type in the `module_param`.

I'd suggest keeping a local copy of `zfs_mmp_interval` which is updated each time through the for loop.  Then use this throughout the loop, that way it can't change mid-loop and you won't need to check it everywhere.  It's good for the whole loop.